src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -1726,6 +1740,15 @@ protected override void InvokeDefaultAction(string path)
                                 return;
                             }
 
+                            //If symlink points to its parent directory then do not pursue symlink further
+                            if (!Platform.IsWindows && Platform.NonWindowsIsSymLink(recursiveDirectory))
```

Comment:
Same here. According to [coding-guideline](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#portable-code), for platform dependent code, we prefer preprocessor directives over runtime checks.